In [1]:
import numpy as np
import pandas as pd
import mysql.connector
from getpass import getpass

In [2]:
password = getpass
password = getpass()

# Dealing with the Marketing Information

### Reading database containing all the information

In [34]:
rd_station = pd.read_csv("../data/lead_complete_rdstation.csv", sep=",")

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\1553664775.py:1: DtypeWarning: Columns (3,4,5,6,7,8,11,12,15,17,24,32,33,35,36,37,40,41,42,43,45,47,48,50,52,53,55,56,57,58,60,63,64,67,69,70,72,73,74,75,76,77,78,80,81,83,85,86,87,88,89,91,92,93,94,95,96,97,100,101,102,103,104,106,108,110,111,112,113,114,115,116,117,118,120,122,123,124,125,126,127,130,131,132,133,134,136,137,138,139,140,141,142,143,144,145,147,148,149,151,152,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,170,172,173,175,176,177,180,181,183,184,186,189,191,192,193,195,196,197,199,200,201,202,203,204,206,207,208,209,210,213,214,215,216,217,218,219,220,222,223,224,225,226,227,229,230,232,234,235,237,239,241,242,244,247,248,250,251,252,253,254,255,256,257,258,260,262,264,265,266,267,268,269,270,271,272,273,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,294,295,296,297,298,300,301,305,306,307,308,309,310,311,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,33

In [35]:
rd_station.columns = rd_station.columns.str.lower().str.replace(' ', '_')

In [36]:
def process_dataframe_columns(df):
    """
    Transform the columns of a DataFrame:
    1. Converts column names to lowercase
    2. Replaces spaces in column names with underscores

    Parameters:
    - DataFrame to be processed.

    Returns:
    - DataFrame with modified column names.
    """
    df.columns = df.columns.str.lower().str.replace(' ', '_')

    return df

rd_station = process_dataframe_columns(rd_station)


In [37]:
rd_station.columns

Index(['email', 'nome', 'telefone', 'celular', 'facebook', 'twitter',
       'linkedin', 'website', 'cargo', 'empresa',
       ...
       'você_tem_interesse_no_curso_de_pós-graduação_em_engenharia_da_manutenção?',
       'você_tem_interesse_no_curso_engenharia_e_gestão_de_ativos?',
       'você_tem_interesse_no_pós-mba_gestão_legal?',
       'você_vai_orientar_um_grupo_de_potenciais_negociadores,_profissionais_de_sua_empresa,_sobre_como_trocar_concessões_(5ª_etapa)._que_recomendações_você_lhe_faria?',
       'você_é_aluno(a)?', 'você_é_convidado_indumak?',
       'você_é_nosso_aluno_regular?', 'você_é_professor(a)_da_sustentare?',
       'área_de_atuação_da_empresa', 'área_de_atuação_de_interesse'],
      dtype='object', length=346)

## Understanding the dataset

In [38]:
rd_station.sample()

email                     nome telefone celular  \
8142  bmarilia.f@malwee.com.br  Marília Bergamo Fogatto      NaN     NaN   

     facebook twitter linkedin website cargo empresa  ...  \
8142      NaN     NaN      NaN     NaN   NaN     NaN  ...   

      você_tem_interesse_no_curso_de_pós-graduação_em_engenharia_da_manutenção?  \
8142                                                NaN                           

     você_tem_interesse_no_curso_engenharia_e_gestão_de_ativos?  \
8142                                                NaN           

     você_tem_interesse_no_pós-mba_gestão_legal?  \
8142                                         NaN   

     você_vai_orientar_um_grupo_de_potenciais_negociadores,_profissionais_de_sua_empresa,_sobre_como_trocar_concessões_(5ª_etapa)._que_recomendações_você_lhe_faria?  \
8142                                                NaN                                                                                                                

     você_é_aluno(a)? você_é_convidado_indumak? você_é_nosso_aluno_regular?  \
8142              NaN                       NaN                         NaN   

     você_é_professor(a)_da_sustentare?  área_de_atuação_da_empresa  \
8142                                NaN                         NaN   

     área_de_atuação_de_interesse  
8142                          NaN  

[1 rows x 346 columns]

In [39]:
rd_station.shape

(41722, 346)

In [40]:
rd_station.isna().sum()

email                                     0
nome                                    698
telefone                              33416
celular                               28981
facebook                              41527
                                      ...  
você_é_convidado_indumak?             41699
você_é_nosso_aluno_regular?           41579
você_é_professor(a)_da_sustentare?    41699
área_de_atuação_da_empresa            41557
área_de_atuação_de_interesse          41722
Length: 346, dtype: int64

In [41]:
rd_station.dtypes

email                                  object
nome                                   object
telefone                               object
celular                                object
facebook                               object
                                       ...   
você_é_convidado_indumak?              object
você_é_nosso_aluno_regular?            object
você_é_professor(a)_da_sustentare?     object
área_de_atuação_da_empresa             object
área_de_atuação_de_interesse          float64
Length: 346, dtype: object

## Dropping the columns that won't need

In [42]:
print(rd_station.columns)

Index(['email', 'nome', 'telefone', 'celular', 'facebook', 'twitter',
       'linkedin', 'website', 'cargo', 'empresa',
       ...
       'você_tem_interesse_no_curso_de_pós-graduação_em_engenharia_da_manutenção?',
       'você_tem_interesse_no_curso_engenharia_e_gestão_de_ativos?',
       'você_tem_interesse_no_pós-mba_gestão_legal?',
       'você_vai_orientar_um_grupo_de_potenciais_negociadores,_profissionais_de_sua_empresa,_sobre_como_trocar_concessões_(5ª_etapa)._que_recomendações_você_lhe_faria?',
       'você_é_aluno(a)?', 'você_é_convidado_indumak?',
       'você_é_nosso_aluno_regular?', 'você_é_professor(a)_da_sustentare?',
       'área_de_atuação_da_empresa', 'área_de_atuação_de_interesse'],
      dtype='object', length=346)


In [43]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(rd_station['cidade'].value_counts())

cidade
Joinville                    2796
Jaraguá do Sul                201
São Paulo                     185
São Bento do Sul              130
Itajaí                         96
Curitiba                       52
Blumenau                       51
Rio de Janeiro                 49
Florianópolis                  41
Belo Horizonte                 30
São Francisco do Sul           28
Fortaleza                      25
Brasília                       25
Sete Lagoas                    21
Navegantes                     20
Vila Velha                     19
Franca                         18
Schroeder                      18
Salvador                       16
Itapoá                         15
Araquari                       15
Goiânia                        14
Campinas                       14
Guaramirim                     14
Recife                         14
Tijucas                        13
Jundiaí                        12
Vitória                        11
Balneário Piçarras             11
Teresin

In [44]:
#rd_station.notnull().sum().sort_values(ascending=False).head(70)
# To be able to look at all the lines
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(rd_station.notnull().sum().sort_values(ascending=False).head(70))

email                                                                                            41722
estágio_no_funil                                                                                 41722
total_de_conversões                                                                              41722
url_pública                                                                                      41722
status_para_comunicação_por_email                                                                41722
lead_scoring_-_interesse                                                                         41722
lead_scoring_-_perfil                                                                            41722
base_legal_para_comunicação                                                                      41660
eventos_(últimos_100)                                                                            41340
origem_da_última_conversão                                               

### If wanted to created a subset with the most responds field

In [45]:
top_50_filled = rd_station.notnull().sum().sort_values(ascending=False).head(50).index
rd_station_columns_responds_most = rd_station[top_50_filled]

## Creating a subset with the information on the lead <> Profile

In [46]:
rd_station_profile = rd_station[[
    'email', 'nome', 'lead_scoring_-_perfil', 'url_pública', 'estágio_no_funil', 'total_de_conversões',
    'lead_scoring_-_interesse', 'status_para_comunicação_por_email',
    'data_da_primeira_conversão', 'eventos_(últimos_100)', 'origem_da_última_conversão',
    'data_da_última_conversão', 'origem_da_primeira_conversão', 'curso_de_interesse:',
    'empresa', 'cargo.1', 'tags', 'data_da_última_oportunidade', 'cargo', 'estado',
    'valor_total_da_oportunidade_no_crm_(última_atualização)',
    'qualificação_da_oportunidade_no_crm_(última_atualização)',
    'etapa_do_funil_de_vendas_no_crm_(última_atualização)',
    'nome_do_responsável_pela_oportunidade_no_crm_(última_atualização)',
    'origem_da_oportunidade_no_crm_(última_atualização)',
    'cidade', 'possui_interesse_em_realizar_uma_pós-graduação_ou_mba?',
    'qual_é_a_sua_área_de_atuação?', 'selecione_o_curso_desejado:.1',
    'qual_a_sua_área_de_atuação?', 'data_de_nascimento',
    'desejo_receber_o_sustentare_news', 'qual_sua_área_de_interesse?',
    'como_você_conheceu_a_sustentare:', 'universidade', 'reside_em:',
    'cidade_de_residência:', 'interesse_em', 'qual_o_curso_de_interesse?'
]]

### Got different columns with representing the same thing, compiling all in one new columns

Represents City: 'Cidade', 'Reside em:' and 'Cidade de residência:'

In [47]:
# Using .loc because was receiving 'SettingWithCopyWarning'
rd_station_profile.loc[:, 'cidade_final'] = (
    rd_station_profile['cidade']
    .fillna(rd_station_profile['reside_em:'])
    .fillna(rd_station_profile['cidade_de_residência:'])
)


C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\1712831611.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'cidade_final'] = (


Represents Position at the company: 'Cargo.1' and 'Cargo'

In [48]:
rd_station_profile.loc[:, 'cargo_final'] = (
    rd_station_profile['cargo.1']
    .fillna(rd_station_profile['cargo'])
)

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\1555167880.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'cargo_final'] = (


Represents area of work in the company: 'Qual a sua área de atuação?', 'Qual a sua área de atuação?'

In [49]:
rd_station_profile.loc[:, 'area_atuacao'] = (
    rd_station_profile['qual_é_a_sua_área_de_atuação?']
    .fillna(rd_station_profile['qual_a_sua_área_de_atuação?'])
)

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\814111575.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'area_atuacao'] = (


Represents course interest: 'Curso de interesse:', 'Qual sua área de interesse?' and 'Interesse em'

In [50]:
rd_station_profile.loc[:, 'interesse_final'] = (
    rd_station_profile['curso_de_interesse:']
    .fillna(rd_station_profile['selecione_o_curso_desejado:.1'])
    .fillna(rd_station_profile['qual_sua_área_de_interesse?'])
    .fillna(rd_station_profile['interesse_em'])
    .fillna(rd_station_profile['qual_o_curso_de_interesse?'])
)

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\2982015989.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rd_station_profile.loc[:, 'interesse_final'] = (


In [51]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(rd_station_profile.isna().sum())

email                                                                    0
nome                                                                   698
lead_scoring_-_perfil                                                    0
url_pública                                                              0
estágio_no_funil                                                         0
total_de_conversões                                                      0
lead_scoring_-_interesse                                                 0
status_para_comunicação_por_email                                        0
data_da_primeira_conversão                                             382
eventos_(últimos_100)                                                  382
origem_da_última_conversão                                             382
data_da_última_conversão                                               382
origem_da_primeira_conversão                                           382
curso_de_interesse:      

With this I was able to reduce 
- *Cidade* had 37078 null values now 36235
- *Cargo* had 27238 null values now 21926
- *Area Atuacao* had 39209 null values now 35800
- *Interesse Final* had 23536 null values now 22223

In [52]:
rd_station_profile = rd_station_profile.drop(columns=['cidade', 'reside_em:', 'cidade_de_residência:', 'cargo.1', 'cargo', 'qual_é_a_sua_área_de_atuação?', 'qual_a_sua_área_de_atuação?', 'curso_de_interesse:', 'qual_sua_área_de_interesse?', 'interesse_em', 'selecione_o_curso_desejado:.1'])

In [53]:
def process_rd_station_profile(df):

    """
    Transform the columns of DataFrame to lowercase and replace spaces in column names with underscores also,
    Process the DataFrame to handle missing values and create consolidated columns.

    Parameters:
    - rd_station (DataFrame): The original RD Station DataFrame.

    Returns:
    - DataFrame: Processed RD Station profile with consolidated and cleaned columns.
        """
    
    df.columns = df.columns.str.lower().str.replace(' ', '_')

    rd_station_profile = df[[
        'email', 'nome', 'lead_scoring_-_perfil', 'url_pública', 'estágio_no_funil', 'total_de_conversões',
        'lead_scoring_-_interesse', 'status_para_comunicação_por_email',
        'data_da_primeira_conversão', 'eventos_(últimos_100)', 'origem_da_última_conversão',
        'data_da_última_conversão', 'origem_da_primeira_conversão', 'curso_de_interesse:',
        'empresa', 'cargo.1', 'tags', 'data_da_última_oportunidade', 'cargo', 'estado',
        'valor_total_da_oportunidade_no_crm_(última_atualização)',
        'qualificação_da_oportunidade_no_crm_(última_atualização)',
        'etapa_do_funil_de_vendas_no_crm_(última_atualização)',
        'nome_do_responsável_pela_oportunidade_no_crm_(última_atualização)',
        'origem_da_oportunidade_no_crm_(última_atualização)',
        'cidade', 'possui_interesse_em_realizar_uma_pós-graduação_ou_mba?',
        'qual_é_a_sua_área_de_atuação?', 'selecione_o_curso_desejado:.1',
        'qual_a_sua_área_de_atuação?', 'data_de_nascimento',
        'desejo_receber_o_sustentare_news', 'qual_sua_área_de_interesse?',
        'como_você_conheceu_a_sustentare:', 'universidade', 'reside_em:',
        'cidade_de_residência:', 'interesse_em', 'qual_o_curso_de_interesse?'
    ]]

    rd_station_profile.loc[:, 'cidade_final'] = (
        rd_station_profile['cidade']
        .fillna(rd_station_profile['reside_em:'])
        .fillna(rd_station_profile['cidade_de_residência:'])
    )

    rd_station_profile.loc[:, 'cargo_final'] = (
        rd_station_profile['cargo.1']
        .fillna(rd_station_profile['cargo'])
    )

    rd_station_profile.loc[:, 'area_atuacao'] = (
        rd_station_profile['qual_é_a_sua_área_de_atuação?']
        .fillna(rd_station_profile['qual_a_sua_área_de_atuação?'])
    )

    rd_station_profile.loc[:, 'interesse_final'] = (
        rd_station_profile['curso_de_interesse:']
        .fillna(rd_station_profile['selecione_o_curso_desejado:.1'])
        .fillna(rd_station_profile['qual_sua_área_de_interesse?'])
        .fillna(rd_station_profile['interesse_em'])
        .fillna(rd_station_profile['qual_o_curso_de_interesse?'])
    )
    rd_station_profile = rd_station_profile.drop(columns=['cidade', 'reside_em:', 'cidade_de_residência:', 'cargo.1', 'cargo', 'qual_é_a_sua_área_de_atuação?', 'qual_a_sua_área_de_atuação?', 'curso_de_interesse:', 'qual_sua_área_de_interesse?', 'interesse_em', 'selecione_o_curso_desejado:.1'])

    return rd_station_profile

# Dealing with the CRM information

### Doing the queries

In [3]:
# Database connection details
db_config = {
    'user': 'um_acesso_bi',
    'password': password,
    'host': '186.250.185.72',
    'database': 'unimestre'
}

def connect_to_database(config):
    try:
        connection = mysql.connector.connect(**config)
        return connection
    except Exception as e:
        print(f"Error connecting to the MySQL database: {e}")
        return None
    
db_connection = connect_to_database(db_config)

In [84]:
def get_student_information(connection):

    
    if connection is None:
        return "Connection to database failed."

    cursor = connection.cursor()
    cursor.execute("SHOW TABLES")
    tables = cursor.fetchall()

    query = f'''SELECT
    pessoas.cd_pessoa,
    moodle.ds_email,
    CONCAT(UPPER(SUBSTRING(pessoas.ds_cidade, 1, 1)), LOWER(SUBSTRING(pessoas.ds_cidade, 2))) as ds_cidade,
    pessoas.dt_nascimento,
    log.min_dt_log
    FROM
    unimestre.pessoas AS pessoas
JOIN 
    unimestre_moodle.view_moodle_pessoas AS moodle
    ON pessoas.cd_pessoa = moodle.cd_pessoa
LEFT JOIN (
    SELECT
        cd_pessoa,
        MIN(dt_log) as min_dt_log
    FROM
        unimestre.diario_logs
    GROUP BY
        cd_pessoa
) AS log ON pessoas.cd_pessoa = log.cd_pessoa
limit 1000000;
'''
    df = pd.read_sql(query, connection)
    df = df.groupby('ds_email').first().reset_index()

    cursor.close()

    return df

info_student_unimestre = get_student_information(db_connection)
info_student_unimestre

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\3170828804.py:31: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


cd_pessoa                  ds_email      ds_cidade  \
0           1001  fabiosilva@unimestre.com       Blumenau   
1           1002        1002a@sememail.com           None   
2         100003        wag_sc@hotmail.com           None   
3         100011  00000000000@sememail.com           None   
4         100046             @sememail.com      São paulo   
...          ...                       ...            ...   
12054  170000461  76835297/00@sememail.com       Blumenau   
12055  170000462  76840198/00@sememail.com  Florianópolis   
12056  170000463   rodrigomnsn@hotmail.com      Joinville   
12057  170000464             @sememail.com       Blumenau   
12058  170000465   170000465a@sememail.com           None   

             dt_nascimento          min_dt_log  
0      2010-01-11 00:00:00 2013-07-26 14:13:39  
1                     None                 NaT  
2      1987-04-02 00:00:00                 NaT  
3      1990-01-01 00:00:00 2013-12-04 10:33:45  
4                     None                 NaT  
...                    ...                 ...  
12054                 None                 NaT  
12055                 None                 NaT  
12056  1991-05-01 00:00:00                 NaT  
12057                 None                 NaT  
12058                 None                 NaT  

[12059 rows x 5 columns]

In [85]:
info_student_unimestre.isna().sum()

cd_pessoa           0
ds_email            0
ds_cidade        2231
dt_nascimento    3431
min_dt_log       8147
dtype: int64

In [86]:
def first_cd_turma_for_students(connection):

    """
    Performs a SQL query to retrieve information about the first 'cd_turma' for students
    from the 'unimestre.view_matriculas' table. Calculates the 'ano' and 'semestre' based on the
    'cd_turma' column and determines the minimum date associated with each 'cd_pessoa', 'cd_turma', 'ano',
    and 'semestre' combination. The result is returned as a Pandas DataFrame.

    Returns:
    - df (DataFrame): A DataFrame containing information about the first 'cd_turma' for students.

    """
        
    if connection is None:
        return "Connection to database failed."

    cursor = connection.cursor()
    cursor.execute("SHOW TABLES")
    tables = cursor.fetchall()

    query = f'''SELECT
    m.cd_pessoa,
    m.cd_turma,
    SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma), 4) AS ano,
    SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma) + 4, 1) AS semestre,
    MIN(
        CASE
            WHEN SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma) + 4, 1) = 'A' THEN DATE_FORMAT(CONCAT(SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma), 4), '-02-01'), '%Y-%m-%d')
            WHEN SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma) + 4, 1) = 'B' THEN DATE_FORMAT(CONCAT(SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma), 4), '-07-01'), '%Y-%m-%d')
            ELSE NULL
        END
    ) AS min_date
FROM
    unimestre.view_matriculas AS m
GROUP BY
    m.cd_pessoa, m.cd_turma, ano, semestre
LIMIT 100000;

# Counting the number of unique cd_pessoa with NULL in min_formatted_date.
SELECT
    COUNT(CASE WHEN min_date IS NULL THEN 1 ELSE NULL END) AS null_min_date_count
FROM (
    SELECT
        m.cd_pessoa,
        MIN(
            CASE
                WHEN SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma) + 4, 1) = 'A' THEN DATE_FORMAT(CONCAT(SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma), 4), '-02-01'), '%Y-%m-%d')
                WHEN SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma) + 4, 1) = 'B' THEN DATE_FORMAT(CONCAT(SUBSTRING(m.cd_turma, LOCATE('20', m.cd_turma), 4), '-07-01'), '%Y-%m-%d')
                ELSE NULL
            END
        ) AS min_date
    FROM
        unimestre.view_matriculas AS m
    GROUP BY
        m.cd_pessoa
) AS subquery
LIMIT 10;

'''
    df = pd.read_sql(query, connection)

    cursor.close()

    return df

first_cd_turma_student = first_cd_turma_for_students(db_connection)

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\1372293948.py:48: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


cd_pessoa       cd_turma   ano semestre    min_date
0         100135   MDINTER2014A  2014        A  2014-02-01
1         100143          CAPAC              A        None
2         100143       MBA2014A  2014        A  2014-02-01
3         100160    ADMOPL2014A  2014        A  2014-02-01
4         100178    MBAOPL2014A  2014        A  2014-02-01
...          ...            ...   ...      ...         ...
15897  170000395   MBAPROJ2013B  2013        B  2013-07-01
15898  170000412    GESTRI2013B  2013        B  2013-07-01
15899  170000430   ENGEMAN2013B  2013        B  2013-07-01
15900  170000430  ENGPROCE2017A  2017        A  2017-02-01
15901  170000430  ENGPROCE2018A  2018        A  2018-02-01

[15902 rows x 5 columns]

In [87]:
first_cd_turma_student.isna().sum()

cd_pessoa       0
cd_turma        0
ano             0
semestre        0
min_date     9162
dtype: int64

In [90]:
# Trying to find the info of the first contact of the student with the school in another table
def first_contact_dates_for_students(connection):

    """ 
    Returns:
    - df (DataFrame): A DataFrame containing information about the first contact date of students.

    SQL query to retrieve information about the first contact date of students
    with the school from the 'unimestre.pessoas' table. The query selects the 'cd_pessoa' and 'dt_cadastro'
    columns, grouping the results by 'cd_pessoa' to get unique records. The result is returned as a DataFrame.
    
    """
    if connection is None:
        return "Connection to database failed."

    cursor = connection.cursor()
    cursor.execute("SHOW TABLES")
    tables = cursor.fetchall()

    query = f'''
    SELECT
        cd_pessoa, dt_cadastro 
    FROM
        unimestre.pessoas
    GROUP BY cd_pessoa
    LIMIT 10; 
'''
    df = pd.read_sql(query, connection)

    cursor.close()

    return df

dt_cadastro_student = first_contact_dates_for_students(db_connection)
dt_cadastro_student

C:\Users\anafl\AppData\Local\Temp\ipykernel_19608\3074074535.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


cd_pessoa dt_cadastro
0           1001  2010-01-01
1           1002         NaT
2         100003  2013-12-02
3         100011  2013-12-03
4         100046  2013-12-04
...          ...         ...
12054  170000461  2013-11-22
12055  170000462  2013-11-22
12056  170000463  2013-11-23
12057  170000464  2013-11-28
12058  170000465         NaT

[12059 rows x 2 columns]

Trying to get less null values on the min_dt_log

In [91]:
merged_df = pd.merge(info_student_unimestre, dt_cadastro_student[['cd_pessoa', 'dt_cadastro']], on='cd_pessoa', how='left')
merged_df = pd.merge(merged_df, first_cd_turma_student[['cd_pessoa', 'min_date']], on='cd_pessoa', how='left')

info_student_unimestre['min_dt_log'] = merged_df['dt_cadastro'].fillna(merged_df['min_date']).fillna(info_student_unimestre['min_dt_log'])

In [92]:
info_student_unimestre.isna().sum()

cd_pessoa           0
ds_email            0
ds_cidade        2231
dt_nascimento    3431
min_dt_log        150
dtype: int64

Was able to reduce 
- *min_dt_log* had 9163 null values now 150

Dropping columns with replicate values

In [99]:
def handle_missing_value_on_min_dt_log(info_student_unimestre, dt_cadastro_student, first_cd_turma_student):
    """
    Merge three DataFrames based on 'cd_pessoa' and handle missing values.

    Parameters:
    - info_student_unimestre (DataFrame): DataFrame containing student information from Unimestre.
    - dt_cadastro_student (DataFrame): DataFrame containing student registration dates.
    - first_cd_turma_student (DataFrame): DataFrame containing the first contact date of students with the school.

    Returns:
    - merged_df (DataFrame): Merged DataFrame with handled missing values.
    """
    # Merge 'info_student_unimestre' with 'dt_cadastro_student'
    merged_df = pd.merge(info_student_unimestre, dt_cadastro_student[['cd_pessoa', 'dt_cadastro']], on='cd_pessoa', how='left')

    # Merge the result with 'first_cd_turma_student'
    merged_df = pd.merge(merged_df, first_cd_turma_student[['cd_pessoa', 'min_date']], on='cd_pessoa', how='left')

    # Handle missing values in 'dt_cadastro'
    info_student_unimestre['min_dt_log'] = merged_df['dt_cadastro'].fillna(merged_df['min_date']).fillna(info_student_unimestre['min_dt_log'])

    return merged_df




# Joining Marketing info with CRM

In [94]:
merge_left = rd_station_profile.merge(info_student_unimestre, left_on='email', right_on='ds_email', how='left')

Creating a new column saying it's a student when a cd_pessoa is found

In [95]:
merge_left['aluno'] = (merge_left['cd_pessoa'].notnull()).astype(int)
merge_left['aluno'].value_counts()

aluno
0    37599
1     4123
Name: count, dtype: int64

## Got different columns with representing the same thing, compiling all in one new columns

In [101]:
def merge_and_preprocess_data(rd_station_profile, info_student_unimestre):
    """
    Merge data from RD Station and Unimestre, and perform necessary preprocessing. Convert into csv

    Parameters:
    - rd_station_profile (DataFrame): DataFrame containing data from RD Station.
    - info_student_unimestre (DataFrame): DataFrame containing student information from Unimestre.

    Returns:
    - merged_df (DataFrame): Merged DataFrame with preprocessed columns.

    """
    # Merge data from RD Station and Unimestre based on the email column
    merge_left = rd_station_profile.merge(info_student_unimestre, left_on='email', right_on='ds_email', how='left')

    # Create a binary column 'aluno' indicating whether a person is a student
    merge_left['aluno'] = (merge_left['cd_pessoa'].notnull()).astype(int)

    # Handle missing values in 'cidade' and 'data_nascimento'
    merge_left.loc[:, 'cidade'] = merge_left['ds_cidade'].fillna(merge_left['cidade_final'])
    merge_left.loc[:, 'data_nascimento'] = merge_left['dt_nascimento'].fillna(merge_left['data_de_nascimento'])

    # Drop unnecessary columns
    merge_left = merge_left.drop(columns=['ds_cidade', 'cidade_final', 'dt_nascimento', 'data_de_nascimento'])

    # Save the result to a CSV file
    merge_left.to_csv("../data/merged_lead_student.csv")

    return merge_left


In [102]:
merge_left = merge_and_preprocess_data(rd_station_profile, info_student_unimestre)

With this I was able to reduce 
- *Cidade* had 35856 null values now 33986
- *Data Nascimento* had 38346 null values now 36117